openqaoa/optimizers/optimization_methods_pennylane.py

```diff
@@ -0,0 +1,160 @@
+
+
+    (read https://docs.pennylane.ai/en/stable/code/api/pennylane.RotosolveOptimizer.html#pennylane.RotosolveOptimizer.step)
+    nums_frequency : dict[dict], required for rotosolve
+        The number of frequencies in the fun per parameter.
+    spectra : dict[dict], required for rotosolve
+        Frequency spectra in the objective_fn per parameter.
+    shifts : dict[dict], required for rotosolve
+        Shift angles for the reconstruction per parameter.
+
+
+    RETURNS
+    -------
+    OptimizeResult : OptimizeResult
+        Scipy OptimizeResult object.
+    '''
+
+    def cost(params, **k): # define a function to convert the params list from pennylane to numpy
+        return fun(np.array(params), *k)
+
+
+    optimizer = AVAILABLE_OPTIMIZERS[method] # define the optimizer
+
+    #get optimizer arguments
+    arguments = inspect.signature(optimizer).parameters.keys()
+    options_keys = list(options.keys())
+
+    #check which values of the options dict can be passed to the optimizer (pop the others)
+    for key in options_keys:
+        if key not in arguments: options.pop(key) 
+        if 'maxiter' in arguments: options['maxiter'] = maxiter
+
+    optimizer = optimizer(**options) #pass the arguments
+    
+    bestx = pl.numpy.array(x0, requires_grad=True)
+    besty = cost(x0, *args)
+    funcalls = 1  # tracks no. of function evals.
+    niter = 0
+    improved = True
+    stop = False  
+
+    testx = np.copy(bestx)
+    testy = np.real(besty)
+    while improved and not stop and niter < maxiter:
+        improved = False
+
+        # compute step
+        if qfim:    #natural_grad_descent
+            testx, testy = optimizer.step_and_cost(cost, bestx, *args, grad_fn=jac, metric_tensor_fn=qfim) 
+        elif jac:   #adagrad, adam, vgd, momentum, nesterov_momentum, rmsprop
+            testx, testy = optimizer.step_and_cost(cost, bestx, *args, grad_fn=jac)
+        elif method=='rotosolve': 
+            testx, testy = optimizer.step_and_cost(
+                                                    cost, bestx, *args,
+                                                    nums_frequency={'params': {(i,):1 for i in range(bestx.size)}} if not nums_frequency else nums_frequency,
+                                                    spectra=spectra,
+                                                    shifts=shifts,
+                                                    full_output=False,
+                                                  )
+        else:       #spsa  
+            testx, testy = optimizer.step_and_cost(cost, bestx, *args)
+
+        # check if stable
+        if np.abs(besty-testy) < tol and niter > 1:
+            improved = False
+
+        else:
+            besty = testy
+            bestx = testx
+            improved = True
+
+        if callback is not None:
+            callback(bestx)
+        if maxfev is not None and funcalls >= maxfev:
+            stop = True
+            break
+
+        niter += 1
```

Comment:
This line was never touched by a test. perhaps the whole function is never tested?